src/main/scala-3/com/raquo/airstream/split/SplittableTypeMacros.scala

```diff
@@ -0,0 +1,220 @@
+package com.raquo.airstream.split
+
+import com.raquo.airstream.core.{EventStream, Signal, Observable, BaseObservable}
+import scala.quoted.{Expr, Quotes, Type}
+import scala.annotation.compileTimeOnly
+
+/**
+ * `SplittableTypeMacros` turns this code
+ * 
+ * ```scala
+ * sealed trait Foo
+ * final case class Bar(strOpt: Option[String]) extends Foo
+ * enum Baz extends Foo {
+ *  case Baz1, Baz2
+ * }
+ * case object Tar extends Foo
+ * val splitter = fooSignal.splitMatch
+ *  .handleCase { case Bar(Some(str)) => str } { (str, strSignal) => renderStrNode(str, strSignal) }
+ *  .handleCase { case baz: Baz => baz } { (baz, bazSignal) => renderBazNode(baz, bazSignal) }
+ *  .handleCase {
+ *    case Tar => ()
+ *    case _: Int => ()
+ *  } { (_, _) => div("Taz") }
+ *  .toSignal
+ * ```
+ * 
+ * into this code:
+ *
+ * ```scala
+ * val splitter = fooSignal.
+ *  .map { i =>
+ *    i match {
+ *      case Bar(Some(str)) => (0, str)
+ *      case baz: Baz => (1, baz)
+ *      case Tar => (2, ())
+ *      case _: Int => (2, ())
+ *    }
+ *  }
+ *  .splitOne(_._1) { ??? }
+ * ```
+ * 
+ * After macros expansion, compiler will warns above code "match may not be exhaustive" and "unreachable case" as expected.
+ */
+object SplittableTypeMacros {
+
+  /**
+   * `MatchSplitObservable` served as macro's data holder for macro expansion.
+   * 
+   * Like any class runtime builder, which uses methods to add various data and a dedicated method to build an instance of the class, `MatchSplitObservable` uses `handleCase`s to keep all code data and `toSignal`/`toStream` to expands macro into to match case definition.
+   * 
+   * For example:
+   *
+   * ```scala
+   *  fooSignal.splitMatch
+   *    .handleCase { case Bar(Some(str)) => str } { (str, strSignal) => renderStrNode(str, strSignal) }
+   *    .handleCase { case baz: Baz => baz } { (baz, bazSignal) => renderBazNode(baz, bazSignal) }
+   * ```
+   * 
+   * will be expanded sematically into:
+   *
+   * ```scala
+   *  MatchSplitObservable.build(fooSignal, ({ case baz: Baz => baz }) :: ({ case Bar(Some(str)) => str }) :: Nil, handlerMap)
+   * ```
+   *
+   * This is important, because `({ case baz: Baz => baz }) :: ({ case Bar(Some(str)) => str }) :: Nil` saves all the infomation needed for the macro to expands into match case definition
+   */
+  final class MatchSplitObservable[Self[+_] <: Observable[_] , I, O] {
+    throw new UnsupportedOperationException("splitMatch without toSignal/toStream is illegal")
+  }
```

Comment:
This class act as a type only. [DuckTape](https://github.com/arainko/ducktape/blob/e001d4a4edb84f64d38c698df6e93ae78cedb706/ducktape/src/main/scala/io/github/arainko/ducktape/BuilderConfig.scala) has a trick to create these kind of type (`opaque TypeName = Unit`), but I cannot seem to integrate it into this code. Will have to dig more into this.

Reply:
Split type definition into a dedicated file seem to work.